code/particle/ParticleSource.h

```diff
@@ -282,6 +282,9 @@ class ParticleSource {
 
 	void initializeThrusterOffset(weapon* wp, weapon_info* wip);
  public:
+	 
+	std::shared_ptr<void> m_effectData; //!< a pointer to a data record specific to the effect type
```

Comment:
I'm sorry but this exactly what I did not want to do with this system since this now puts particle source data in random locations in memory. Instead, you could use a union (although that is a hassle to work with in C++) or a variant (via https://github.com/mpark/variant) to store effect specific data directly in the source.

Reply:
I think a union should be ok for now, since we can use the `particleEffect` as the descriminant. And if locality is the goal, I assume that a vector (of spouts, what 'spout' types use `m_effectData` for) is probably out too? Using an array and a maximum number of spouts per source isn't the worst thing in the world, but obviously it would be nice to keep the unboundedness.

Reply:
Just out of curiosity any update on a consensus @asarium?

Reply:
I'll need to change it to a union and with an array for the spouts, I think. I've just left this one hanging for awhile.

Reply:
Another option might be to just use the actions system for this. You could have actions that create single effects that would then act like the "spouts" this wants to implement.